File: webapp-net/Ugc/Controllers/UgcController.cs
using System.Collections.Generic;
using System.Linq;
using Sdl.Web.Mvc.Controllers;
using Sdl.Web.Common.Models;
using Sdl.Web.Modules.Ugc.Data;
using Sdl.Web.Modules.Ugc.Models;
using Sdl.Web.Mvc.Configuration;

namespace Sdl.Web.Modules.Ugc.Controllers
{
    /// <summary>
    /// Ugc Controller
    /// </summary>
    public class UgcController : EntityController
    {
        protected override ViewModel EnrichModel(ViewModel sourceModel)
        {
            UgcComments model = base.EnrichModel(sourceModel) as UgcComments;
            if (model != null)
            {
                if (Request.HttpMethod == "POST")
                {
                    // don't retrieve comments on a POST and to keep rendering happy give it an
                    // empty list
                    model.Comments = new List<UgcComment>();
                }
                else
                {
                    var ugcService = new UgcService();
                    var comments = ugcService.GetComments(model.Target.PublicationId, model.Target.ItemId, false,
                        new int[] {}, 0, 0);
                    model.Comments = CreateEntities(comments);
                }

                return model;
            }

            UgcPostCommentForm postForm = base.EnrichModel(sourceModel) as UgcPostCommentForm;
            if (postForm != null && MapRequestFormData(postForm) && ModelState.IsValid)
            {
                var ugcService = new UgcService();
                ugcService.PostComment(postForm.Target.PublicationId, postForm.Target.ItemId, postForm.UserName, postForm.EmailAddress, postForm.Content, postForm.ParentId, postForm.Metadata);
                return new RedirectModel(WebRequestContext.RequestUrl);
            }

            return sourceModel;
        }

        private static List<UgcComment> CreateEntities(List<Comment> comments) => comments.Select(CreateEntity).ToList();

        private static UgcComment CreateEntity(Comment comment) => new UgcComment {Comments = CreateEntities(comment.Children), CommentData = comment};
    }
}

File: webapp-java/dxa-module-ugc/src/main/java/com/sdl/dxa/modules/ugc/controllers/UgcController.java
package com.sdl.dxa.modules.ugc.controllers;

import com.sdl.dxa.modules.ugc.UgcService;
import com.sdl.dxa.modules.ugc.data.Comment;
import com.sdl.dxa.modules.ugc.model.entity.UgcComment;
import com.sdl.dxa.modules.ugc.model.entity.UgcComments;
import com.sdl.dxa.modules.ugc.model.entity.UgcPostCommentForm;
import com.sdl.webapp.common.api.model.EntityModel;
import com.sdl.webapp.common.api.model.ViewModel;
import com.sdl.webapp.common.controller.EntityController;
import org.springframework.stereotype.Controller;

import javax.servlet.http.HttpServletRequest;
import java.util.ArrayList;
import java.util.List;

@Controller
public class UgcController extends EntityController {

    private static List<UgcComment> createEntities(List<Comment> comments) {

        final List<UgcComment> ugcComments = new ArrayList<>();
        comments.forEach((Comment comment) -> ugcComments.add(createEntity(comment)));
        return ugcComments;
    }

    private static UgcComment createEntity(Comment comment) {
        final UgcComment ugcComment = new UgcComment();
        ugcComment.setComments(createEntities(comment.getChildren()));
        ugcComment.setCommentData(comment);
        return ugcComment;
    }

    @Override
    protected ViewModel enrichModel(ViewModel model, HttpServletRequest request) throws Exception {

        if (model instanceof UgcComments) {
            final ViewModel enrichedModel = super.enrichModel(model, request);
            final UgcService ugcService = new UgcService(context);
            final UgcComments ugcComments = (UgcComments) (enrichedModel instanceof EntityModel ? enrichedModel : model);
            final List<Comment> comments = ugcService.getComments(ugcComments.getTarget().getPublicationId(),
                    ugcComments.getTarget().getItemId(), false, new Integer[0], 0, 0);
            ugcComments.setComments(createEntities(comments));
            return ugcComments;
        }

        if (model instanceof UgcPostCommentForm) {
            final ViewModel enrichedModel = super.enrichModel(model, request);
            final UgcPostCommentForm postForm = (UgcPostCommentForm) (enrichedModel instanceof EntityModel ? enrichedModel : model);
            postForm.setFormUrl(context.getPage().getUrl());

        }
        return model;
    }
}
